src/Compilers/CSharp/Portable/Binder/Binder_Lookup.cs

```diff
@@ -1322,8 +1322,12 @@ internal SingleLookupResult CheckViability(Symbol symbol, int arity, LookupOptio
                 ? ((AliasSymbol)symbol).GetAliasTarget(basesBeingResolved)
                 : symbol;
 
+            if (unwrappedSymbol is SourceNamedTypeSymbol { IsFile: true } && !unwrappedSymbol.IsDefinedInSourceTree(this.AssociatedSyntaxTree, definedWithinSpan: null))
```

Comment:
I'd like to see how the lookup rules are spec'ed before reviewing this implementation.

Reply:
For instance, should we instead introduce a binder before the containing type that would introduce the file-types into scope? It would be good to check with @AlekseyTs, but my feeling is that adding a "file" binder in the chain will help with semantic APIs later on (LookupSymbol and such).

Reply:
> There's a question of whether file-types are visible-but-inaccessible from another file, or whether they are invisible from another file.

IMO, it's sufficient for the specification to say that file types are not *visible* from other files. For now, we behave as though the file type doesn't exist when looking up in such a scenario.

For diagnostic quality we could consider giving a message a la "*CLASS is inaccessible due to its protection level*". It would be fairly simple to accomplish, I think, since LookupResult is already geared toward determining and surfacing the *best* diagnostic for a failed lookup. But I don't think it's necessary to do this as part of the core feature implementation.